server/src/main/java/org/apache/druid/server/ClientQuerySegmentWalker.java

```diff
@@ -429,22 +432,18 @@ private DataSource inlineIfNecessary(
       } else if (canRunQueryUsingLocalWalker(subQuery) || canRunQueryUsingClusterWalker(subQuery)) {
         // Subquery needs to be inlined. Assign it a subquery id and run it.
 
-        final Sequence<?> queryResults;
+        final Function<Query<T>, QueryRunner<T>> queryRunnerSupplier;
 
         if (dryRun) {
-          queryResults = Sequences.empty();
+          queryRunnerSupplier = query -> (queryPlus, responseContext) -> Sequences.empty();
         } else {
-          final QueryRunner subqueryRunner = subQuery.getRunner(this);
-          queryResults = subqueryRunner.run(
-              QueryPlus.wrap(subQuery),
-              DirectDruidClient.makeResponseContextForQuery()
-          );
+          queryRunnerSupplier = query -> query.getRunner(this);
```

Comment:
I'm not sure you need this to be a `Function<>`.  I think it can just be `QueryRunner<T>`.  It should be safe to get the QueryRunner here and then just pass in the query with the context when you actually run it.

That, or you can make the decision on what serialization format you are going to ask for ahead of this and add it to the context first.  Then you probably don't even have to change any of the implementation below.

Reply:
The reason for it being a function was that there's a `QuerySwappingRunner` which ensures that the query with which the runner was created is the same as the query with which it is being run and modifying the context after creating the runner [throws with exception](https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/server/QuerySwappingQueryRunner.java#L57-L57). 

The latter suggestion works, and I have modified the code to that effect.